_posts/2022-03-08-stream-processing-part2.md

```diff
@@ -0,0 +1,518 @@
+
+  results.withTimestamp should inOnTimePane("00:00:00", "00:11:00") {
+    containSingleValueAtTime("00:10:59.999", ("joe", 
+      Iterable("open app", "close app")))
+  }
+}
+```
+
+* The activities in the session are ordered by activity event time
+* The result element gets end-of-window time, ten minutes after the oldest event time 
+* The oldest event do not have to be necessarily the latest one
+
+## Continues long-lasting session
+
+Let's check if the continuous session is preserved for activities that span for a longer period of time but still within the allowed gap.
+
+```scala
+"Continuous activities" should "be aggregated into single session" in runWithContext { sc =>
+  val activities = testStreamOf[(User, Activity)]
+    .addElementsAtTime("00:00:00", ("joe", "open app"))
+    .addElementsAtTime("00:01:30", ("joe", "show product"))
+    .addElementsAtTime("00:03:00", ("joe", "add to cart"))
+    .addElementsAtTime("00:09:30", ("joe", "checkout"))
+    .addElementsAtTime("00:13:10", ("joe", "close app"))
+    .advanceWatermarkToInfinity()
+
+  val results = activitiesInSessionWindow(sc.testStream(activities), TenMinutesGap)
+
+  results.withTimestamp should inOnTimePane("00:00:00", "00:23:10") {
+    containSingleValueAtTime(
+      "00:23:09.999",
+      ("joe", Iterable("open app", "show product", "add to cart", "checkout", "close app")))
+  }
+}
+```
+
+* There is no gap longer than 10 minutes in the activities stream
+* As expected, the single, continuous session is produced
+* Result element gets end-of-window time *00:23:09.999*, ten minutes more than the oldest observed event time
+
+## Interrupted long-lasting session
+
+What if the gap between events is longer than ten minutes?
+
+```scala
+"Interrupted activities" should "be aggregated into two sessions" in runWithContext { sc =>
+  val activities = testStreamOf[(User, Activity)]
+    .addElementsAtTime("00:00:00", ("joe", "open app")
+    .addElementsAtTime("00:01:30", ("joe", "show product"))
+    .addElementsAtTime("00:03:00", ("joe", "add to cart"))
+    .addElementsAtTime("00:13:00", ("joe", "checkout"))
+    .addElementsAtTime("00:13:10", ("joe", "close app"))
+    .advanceWatermarkToInfinity()
+
+  val results = activitiesInSessionWindow(sc.testStream(activities), TenMinutesGap)
+
+  results.withTimestamp should inOnTimePane("00:00:00", "00:13:00") {
+    containSingleValueAtTime("00:12:59.999",
+      ("joe", Iterable("open app", "show product", "add to cart")))
+  }
+
+  results.withTimestamp should inOnTimePane("00:13:00", "00:23:10") {
+    containSingleValueAtTime("00:23:09.999", 
+      ("joe", Iterable("checkout", "close app")))
+  }
+}
+```
+
+* The gap between "add to cart" and "checkout" is too long to produce the single session
+* Two independent sessions are produced
+* Result element always gets end-of-window time, ten minutes more than the oldest observed activity for the given session
+
+## Late activity
```

Comment:
W tym scenariusze brakuje mi trochę jawnego stwierdzenia, że dozwolone opóźnienie zdarzeń jest równe "0".

Sygnatura metody na początku artykułu tego nie zawiera (w sumie słusznie - żeby nie zaciemniać), ale może warto by przed przejściem do testu wtrącić zdanie, że rozpatrujemy przypadek z `allowedLateness = ZERO`?

Reply:
Słuszna uwaga, pomyślę jako to ograć. Analogicznie nie omawiam też jawnie domyślnego trigger-a, a zapewne warto.